x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/sender/TimedListener.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.http.sender;
+
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ListenerTimeouts;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.xpack.inference.InferencePlugin.UTILITY_THREAD_POOL_NAME;
+
+/**
+ * Provides a way to set a timeout on the listener. If the time expires, the original listener's
+ * {@link ActionListener#onFailure(Exception)} is called with an error indicating there was a timeout.
+ *
+ * @param <Response> the type of the value that is passed in {@link ActionListener#onResponse(Object)}
+ */
+public class TimedListener<Response> {
+
+    private final ActionListener<Response> listenerWithTimeout;
+    private final AtomicBoolean finished = new AtomicBoolean();
```

Comment:
```suggestion
    private final AtomicBoolean completed = new AtomicBoolean();
```

nit: the getter is called `hasCompleted`, maybe we should give the variable the same name